### PR TITLE
Implement bi-directionality

### PR DESCRIPTION
**Pull request by yair-schiff**
- [x] Implement bi-directionality by applying backbone twice: (1) to the forward sequence and (2) to the backward sequence.
- [x] Implement 3 strategies for combining forward / backward Mamba hidden states:
    1. `add`: Add the states.
    2. `concat`: Concatenate the states. This doubles the hidden dimension,`d_model`, which also prevents weight tying between `embedding` and `lm_head` weights.
    3. `ew_multiply`: perform element-wise multiplication between the states.

### CODE REVIEW

File: mamba_ssm/models/mixer_seq_simple.py [hunk 1]
```diff
@@ -83,6 +90,95 @@ def _init_weights(
...
+        hidden_states: (B, L, D)
+        Returns: same shape as hidden_states
+        """
+        out = self.mamba_fwd(hidden_states, inference_params=inference_params)
+        if self.bidirectional:
+            out_rev = self.mamba_rev(
+                hidden_states.flip(dims=(1,)),  # Flip along the sequence length dimension
+                inference_params=inference_params
+            ).flip(dims=(1,))  # Flip back for combining with forward hidden states
+            if self.bidirectional_strategy == "add":
```
Comment (Skylion007) [inline mamba_ssm/models/mixer_seq_simple.py:175]: bidirectional strategy should be an enum

File: mamba_ssm/models/mixer_seq_simple.py [hunk 2]
```diff
@@ -234,8 +338,8 @@ def forward(self, input_ids, position_ids=None, inference_params=None, num_last_
         if num_last_tokens > 0:
             hidden_states = hidden_states[:, -num_last_tokens:]
         lm_logits = self.lm_head(hidden_states)
-        CausalLMOutput = namedtuple("CausalLMOutput", ["logits"])
-        return CausalLMOutput(logits=lm_logits)
+        LMOutput = namedtuple("LMOutput", ["logits"])
```
Comment (Skylion007) [inline mamba_ssm/models/mixer_seq_simple.py:341]: I know this isn't from your code, but constructing named tuples in this way is deprecated as it prevents proper type inference.

**Review by Skylion007 — COMMENTED** [review]
Left some nits

**Comment by sentialx** [comment]
What if the sequences have paddings? E.g.
Input is
[1 2 3 0 0 0]
So flipped input would be
[0 0 0 3 2 1].
Shouldn't it be
[3 2 1 0 0 0]?

**Comment by yair-schiff** [comment]
@sentialx , agreed. That's a good catch.

**Comment by jimmieliu** [comment]
how the speed compares to uni-directional?

**Comment by yair-schiff** [comment]
> how the speed compares to uni-directional?

It's about 2x

**Comment by pengzhangzhi** [comment]
@yair-schiff I am just curious, did you solve the 

> What if the sequences have paddings? E.g. Input is [1 2 3 0 0 0] So flipped input would be [0 0 0 3 2 1]. Shouldn't it be [3 2 1 0 0 0]?

Just curious, is this problem solved?

**Comment by pengzhangzhi** [comment]
I came up with a solution to the padding issue. Say a tensor [1,2,3,0,0], where 0 is the padding token. We flip it to get [0,0,1,2,3], pass it to the network and flip it back. Therefore, the flipped tensor information matches the original tensor order as we apply double flips.
```
given: x
out = x + f(x.flip()).flip()
```

**Comment by xuanwuji** [comment]
> I came up with a solution to the padding issue. Say a tensor [1,2,3,0,0], where 0 is the padding token. We flip it to get [0,0,1,2,3], pass it to the network and flip it back. Therefore, the flipped tensor information matches the original tensor order as we apply double flips.
> 
> ```
> given: x
> out = x + f(x.flip()).flip()
> ```

Hi,  Your approach is clever! But I have a question: if you flip the input to [0,0,1,2,3], does the padding in front of it affect sequence hidden features learning? i.e., does it produce a different result(bad repersentation of sequence) than the input of [3,2,1,0,0]?
I don't know enough about it, could you possibly give me some guidance? This will help me a lot. Thank you very much!

**Comment by Museum7432** [comment]
@xuanwuji well, you can remove the leading paddings by shifting each row of x before flipping x. As for its effect, since the hidden state is initialized with 0, it should still be filled with 0 after scanning through the paddings. So, those padding shouldn't have any effect on the result. However, you can use the following function just to be sure.
```
def flip_padded_hidden_states(hidden_states, seq_lens):
    batch_size, seq_len, hidden_dim = hidden_states.shape

    indices = torch.arange(batch_size * seq_len, device=hidden_states.device).reshape(
        batch_size, seq_len
    )

    indices_offset = seq_len - seq_lens

    indices = (indices - indices_offset.unsqueeze(1)) % (seq_len * batch_size)

    indices = indices.flip(1)

    return hidden_states.reshape(batch_size * seq_len, hidden_dim)[indices]
```
checking the effect of paddings:
```
import torch
from mamba_ssm import Mamba2, Mamba
from torch.nn import functional as F

batch, length, dim = 2, 64, 16

model = Mamba(
    d_model=dim, # Model dimension d_model
    d_state=16,  # SSM state expansion factor
    d_conv=4,    # Local convolution width
    expand=2,    # Block expansion factor
).to("cuda")

x = torch.randn(batch, length, dim).to("cuda")
padded_x = F.pad(x, (0,0, 4,0))

y = model(x)
padded_y = model(padded_x)

unpadded_y = padded_y[:,4:]

print(f'Output max diff: {(unpadded_y - y).abs().max().item()}')
print(f'Output mean diff: {(unpadded_y - y).abs().mean().item()}')
```
